src/server/search/doc_index.cc

```diff
@@ -56,13 +56,35 @@ const absl::flat_hash_map<string_view, search::SchemaField::FieldType> kSchemaTy
     {"NUMERIC"sv, search::SchemaField::NUMERIC},
     {"VECTOR"sv, search::SchemaField::VECTOR}};
 
+size_t GetProbabilisticBound(size_t hits, size_t requested, optional<search::AggregationInfo> agg) {
+  auto intlog2 = [](size_t x) { return int(log2(x)); };  // TODO: replace with loop or builting_clz
+  size_t shards = shard_set->size();
+
+  // Estimate how much every shard has with at least 99% prob
+  size_t avg_shard_min = hits * intlog2(hits) / (12 + shard_set->size() / 10);
+  avg_shard_min -= min(avg_shard_min, min(hits, size_t(5)));
+
+  // If it turns out that we might have not enough results to cover the request, don't skip any
+  if (avg_shard_min * shards < requested)
+    return requested;
+
+  // If all shards have at least avg min, keep the bare minimum needed to cover the request
+  size_t limit = requested / shards + 1;
+
+  // Aggregations like SORTBY and KNN reorder the result and thus introduce some variance
```

Comment:
Don't we have to fetch all matching for SORTBY?

Reply:
No/Yes :nerd: We have to fetch all, but not serialize all. That is what the doc reference is for. If we find out when building the reply that some entries should have been included but are not serialized, we'll refill (see BuildSortedOrder())